### DISCUSSION TITLE
请问有没有什么方法实现类似UnLuaInterface的效果，既蓝图实例的生命周期去控制一个脚本对象？

### DISCUSSION THREAD

**Discussion by watsonsong** [Tencent/puerts#1214] [Q&A]
引擎对象控制脚本对象的生命周期是一个挺好用的功能，但我看了继承引擎类，和Mixin似乎都做不到。
能否有类似Unlua中[UnLuaInterface](https://github.com/Tencent/UnLua/blob/master/Plugins/UnLua/Source/UnLua/Public/UnLuaInterface.h)的能力，蓝图可以实现这个接口，返回脚本上的类名，会自动mixin到这个对象？

**Comment by chexiongsheng** [comment]
UnLuaInterface其实就起一个作用，配置一个字符串形式的lua模块名。
这在动态类型语言可以这么搞。
静态类型如何保证你配的那个模块名和那个蓝图的接口是匹配的呢？靠人工肉眼保证么？当然可以不管，当动态类型语言来，但这已经废掉静态语言的最大好处了。

**Comment by chexiongsheng** [comment]
其次这种是对业务代码的侵入。
最后还有我个人认为不适应的蓝图资源（美术的产出）对代码（程序产出）的反向依赖。

**Comment by chexiongsheng** [comment]
提供这功能当然及其简单
只是并不合适，引入的问题多过解决的问题。

ps，用一项技术就应按那项技术的思维方式来思考。有时你觉得的不便，只是你的思维方式错了。

**Comment by watsonsong** [comment]
对静态语言也有一些办法，比如通过标记之类导出一个字符串的工厂函数。
引擎对象持有脚本对象的生命周期是一个比较好用的行为，比如UI界面的WBP，内部有一些组件多层嵌套，或者一个ListView的Entry上希望挂的脚本逻辑。可能是我没有找到好的方法，希望伴随引擎的创建对象自动把关联的脚本创建出来。而不是实例化了UI面板之后，再遍历去挂（Mixin）各个子对象的脚本。
如果扩展一下UUserWidget上去绑定一个TS对象似乎也可以，但如果是UnLuaInterface这种去Hook了UObject的创建，其实就支持了任意蓝图对象更加方便。

**Comment by chexiongsheng** [comment]
不明白你说的字符串的工厂函数是啥意思？我说的静态类型检查，是编译期的事情，而且不仅仅是检查个类型名字，还有接口是否匹配，比如你蓝图定义了foo(a:number):void接口，你的ts里头定义的是foo(a:string):void会报错，你是希望在运行时来一次和编译差不多开销的静态类型检查么。
你说“引擎对象持有脚本对象的生命周期”，意思有点偏差，总让人往类实例生命周期想。其实UnLuaInterface没啥，就一个配置，配置了要绑定到哪个lua模块。

你要做类似的事情完全可以在现有puerts提供的功能实现，搞些公用基类即可。

**Comment by watsonsong** [comment]
例如创建了一个WBP的蓝图，其中放置了一些UI控件并且导出成变量。这个时候生成的ts定义文件中就会有这个BP的类型检查的定义了。
要使用这个UI的话，需要在TS中用UE.Class.Load起来，转换成对应的类图类，其中的变量等都可以访问。

但如果这个WBP并不是由脚本直接创建，而是ListView或者嵌套在其他UI里面的一个子控件的时候，就希望对应的ts类能自动的创建出来。例如UnLuaInterface制定的脚本模块，里面能直接访问WBP导出的各个变量。

可能对UE和puerts都并不是太熟悉，想请教一下如果希望一个WBP或者Actor在实例化的时候，能伴随创建一个ts类的对象，同时把这个BP对象传递给ts，因为BP对应自动也有ts的签名，所以BP本身的变量和函数应该都能访问到。如果这样的话，公用基类之类的怎么组织会好一点呢？

**Reply by chexiongsheng** [reply]
UnLuaInterface那是个字符串，填啥都行。你怎么验证这个字符串所指向ts模块和蓝图的接口是匹配的呢？或者更实际的例子，你名字输错成另外的一个模块怎么办？

**Reply by watsonsong** [reply]
例如TS每个class上给一个标记，然后设计一个GetBindClasses的接口，返回所有能绑定的类名字和构造函数。
编辑器模式下也启动TS实例调用这个接口，上面可以调用一下ts实例获取这个列表，表现成下拉选项框，或者字符串写错了显示成红色之类的。

**Reply by chexiongsheng** [reply]
什么接口可以运行时返回ts类型信息？ts编译后类型信息就擦除了

**Reply by watsonsong** [reply]
meta自己写一个后处理生成一些代码大概如此~

**Comment by chexiongsheng** [comment]
不介意没静态类型校验丢失的话。实现同等功能就几十行代码：
1、通过FUObjectArray::FUObjectCreateListener监听UE对象创建；
2、发现是继承UnLuaInterface的CDO，获取js模块名，然后调用绑定到std::function<UClass*, FString>（或者Delegate）的js函数，传递UClass和绑定的js模块名；
3、在2里头被调用js函数调用mixin；

**Reply by watsonsong** [reply]
感谢感谢，我参考着自己实现试试

**Reply by ljporljp** [reply]
mixin一定要在js端执行么

**Reply by chexiongsheng** [reply]
不一定

**Reply by chexiongsheng** [reply]
补充下，这种方式就是unlua的方式，据项目反馈这种方式在多线程下有些奇怪问题。推荐我最后提供方案，某项目用了很久没什么问题。

**Comment by chexiongsheng** [comment]
> 不介意没静态类型校验丢失的话。实现同等功能就几十行代码： 1、通过FUObjectArray::FUObjectCreateListener监听UE对象创建； 2、发现是继承UnLuaInterface的CDO，获取js模块名，然后调用绑定到std::function<UClass*, FString>（或者Delegate）的js函数，传递UClass和绑定的js模块名； 3、在2里头被调用js函数调用mixin；

补充一下，这种方式不太建议，实际使用中，反射调用Interface在UObject创建回调中不可用。

某项目的自动mixin的实践是：是需要用到的类，比如Actor，搞个c++基类，覆盖OnCreate，在OnCreate自动mixin，然后蓝图需要继承Actor的都改为继承这个。

**Reply by kong21690** [reply]
车神 有完整案例给一下吗？感觉这样方便一些

**Reply by chexiongsheng** [reply]
> 车神 有完整案例给一下吗？感觉这样方便

不是我们项目，如果某天那项目看到这讨论，他们又愿意分享。看你们的缘分。
不过描述了思路你们自己来应该也不难。

**Reply by kong21690** [reply]
好滴 车神回复好快，我去看下c++ mixi的代码
